packages/vscode-extension/src/webview/components/PreviewLoader.tsx

```diff
@@ -25,23 +35,44 @@ function PreviewLoader({ onRequestShowPreview }: { onRequestShowPreview: () => v
         (item) => item.StartupMessage === projectState.startupMessage
       );
       const currentWeight = StartupStageWeight[currentIndex].weight;
-      const startupStageWeightSumUntillNow = StartupStageWeight.slice(0, currentIndex)
+      const startupStageWeightSumUntilNow = StartupStageWeight.slice(0, currentIndex)
         .map((item) => item.weight)
-        .reduce((acc, cur) => (acc += cur), 0);
+        .reduce((acc, cur) => acc + cur, 0);
 
       let progressComponent = 0;
 
       if (projectState.stageProgress !== undefined) {
         progressComponent = projectState.stageProgress;
       }
       setProgress(
-        ((startupStageWeightSumUntillNow + progressComponent * currentWeight) /
+        ((startupStageWeightSumUntilNow + progressComponent * currentWeight) /
           startupStageWeightSum) *
           100
       );
     }
   }, [projectState]);
 
+  useEffect(() => {
+    const { status, startupMessage } = projectState;
+    if (previousProgress.status !== status) {
+      clearTimeout(isLoadingSlowlyTimeout.current);
+    }
+    if (previousProgress.startupMessage !== startupMessage) {
+      clearTimeout(isLoadingSlowlyTimeout.current);
+      // We skip reporting slow builds, this is the only most time-consuming
+      // task and times varies from project to project.
+      if (startupMessage !== StartupMessage.Building) {
+        isLoadingSlowlyTimeout.current = setTimeout(() => {
+          setIsLoadingSlowly(true);
+        }, slowLoadingThresholdMs);
+      }
+    }
+    setPreviousProgress({
+      status: projectState.status,
+      startupMessage: projectState.startupMessage,
+    });
+  }, [projectState]);
```

Comment:
shouldn't we just have projectState.status on deps list and then we don't need to add previousStatus as new state because useEffect will handle this for us